### ISSUE TITLE
signald group v2 issue

### ISSUE THREAD

**Issue by demure** [hoehermann/purple-signald#26]
My one group chat recently updated to a v2 group, add it looks like libpurple-signald may be having issues handling v2 groups.
Currently  in my setup involving bitlbee:
* Content I post to the group goes to a /privmsg named '-sig' (as I set bitlebee to append '-sig' to buddy names for signal)
* Content my friend posts to the group shows up in in the /privmsg for thier buddy name as if it was one-on-one chat

Examples from NC:
```
## Them posting to group v2
{
  "type": "message",
  "data": {
    "username": "+ME",
    "source": {
      "number": "+THEM",
      "uuid": "REDACTED"
    },
    "sourceDevice": 1,
    "type": "CIPHERTEXT",
    "timestamp": 1610048610205,
    "timestampISO": "2021-01-07T19:43:30.205Z",
    "serverTimestamp": 1610048609015,
    "serverDeliveredTimestamp": 1610048609019,
    "hasContent": true,
    "dataMessage": {
      "timestamp": 1610048610205,
      "body": "Test received",
      "groupV2": {
        "id": "REDACTED",
        "revision": 1
      }
    }
  }
}

## Me to group v2
{
  "type": "message",
  "data": {
    "username": "+ME",
    "source": {
      "number": "+ME",
      "uuid": "REDACTED"
    },
    "sourceDevice": 1,
    "type": "CIPHERTEXT",
    "timestamp": 1610082077851,
    "timestampISO": "2021-01-08T05:01:17.851Z",
    "serverTimestamp": 1610082080136,
    "serverDeliveredTimestamp": 1610082080137,
    "hasContent": true,
    "syncMessage": {
      "sent": {
        "timestamp": 1610082077851,
        "message": {
          "timestamp": 1610082077851,
          "body": "Unmute for rapid fire... https://imgur.com/gallery/qslXPb7",
          "groupV2": {
            "id": "REDACTED",
            "revision": 1
          },
          "previews": [
            {
              "url": "https://imgur.com/gallery/qslXPb7",
              "title": "Unmute for rapid fire...",
              "attachment": {
                "contentType": "image/jpeg",
                "id": "REDACTED",
                "size": 28634,
                "width": 600,
                "height": 315,
                "key": "REDACTED",
                "digest": "REDACTED",
                "blurhash": "REDACTED"
              }
            }
          ]
        },
        "unidentifiedStatus": {
          "+THEM": true,
          "REDACTED": true
        }
      }
    }
  }
}
```

**Comment by hoehermann** [comment]
As of today, this plug-in does not support the `groupV2` format. If I create a group with my signald test account being one of the members, it defaults to an old-style group, so I have no way to test this. :( 

I am not quite sure how to accept invitations to groups. Though I expect Finn to answer [my question](https://gitlab.com/signald/signald/-/merge_requests/31#note_481550776) soon.

Since I did a lot of experiments regarding registering and linking lately, all my remaining numbers seem to be blocked by Signal.

This is an issue to be resolved later, I guess.

**Comment by hoehermann** [comment]
I was able to re-enable group chats to some extent. I can receive messages, but the new format uses uuids rather than the numbers to identify group members. This needs additional work. @fancypantalons are you interested in picking this up?

PS: I thought you redacted stuff when you wrote "CIPHERTEXT". But it actually says CIPHERTEXT verbatim. XD

**Comment by Strykar** [comment]
@hoehermann
Unsure if I should create a new issue for this..
Today's two commits, [a493c6d](https://github.com/hoehermann/libpurple-signald/commit/a493c6df8ffe36fda545653ecbda7e47eef89a4a) and [8240f34](https://github.com/hoehermann/libpurple-signald/commit/8240f34008e2295f890df9966b06f4c47bf6c623), cause the build to fail with:
```
cc -shared -o libsignald.so comms.o contacts.o direct.o groups.o link.o message.o libsignald.o -Wl,-O1,--sort-common,--as-needed,-z,relro,-z,now 
pkg-config purple glib-2.0 json-glib-1.0 --libs

/usr/bin/ld: libsignald.o: warning: relocation against 
signald_login' in read-only section
.text'
/usr/bin/ld: contacts.o: relocation R_X86_64_PC32 against symbol `signald_process_contact' can not be used when making a shared object; recompile with -fPIC
/usr/bin/ld: final link failed: bad value
collect2: error: ld returned 1 exit status
make: *** [Makefile:51: libsignald.so] Error 1
==> ERROR: A failure occurred in build().
    Aborting...
```
Tested on `Linux bitlbee_host 5.10.7-arch1-1 #1 SMP PREEMPT Wed, 13 Jan 2021 12:02:01 +0000 x86_64 GNU/Linux`

**Comment by demure** [comment]
the most recent commit of a493c6d compiles and runs fine for me on debian sid, and has returned my group to working.

**Comment by hoehermann** [comment]
@Strykar Looks like I messed up the Makefile. The old ones should continue to work.

**Comment by Strykar** [comment]
> @Strykar Looks like I messed up the Makefile. The old ones should continue to work.

That still fails, just further along - https://gist.github.com/Strykar/7690fa00754d1ce5f2c367d626003893

**Comment by hoehermann** [comment]
For some reasons, your build system drops the -fPIC option, which was indeed considered optional. I changed the Makefile so it is mandatory.

**Comment by Strykar** [comment]
It now fails at:
```
/usr/include/libpurple/media/../util.h:1397:13: warning: Deprecated pre-processor symbol, replace with 
 1397 | G_CONST_RETURN gchar *purple_gai_strerror(gint errnum);
      |             ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~     
ld -shared -o libsignald.so comms.o contacts.o direct.o groups.o link.o message.o libsignald.o -Wl,-O1,--sort-common,--as-needed,-z,relro,-z,now `pkg-config purple glib-2.0 json-glib-1.0 --libs`
ld: unrecognized option '-Wl,-O1,--sort-common,--as-needed,-z,relro,-z,now'
ld: use the --help option for usage information
make: *** [Makefile:52: libsignald.so] Error 1
==> ERROR: A failure occurred in build().
    Aborting...
```

**Comment by hoehermann** [comment]
What compilers and linkers are you using?

**Comment by Strykar** [comment]
@hoehermann 
```
$ gcc -v
Using built-in specs.
COLLECT_GCC=gcc
COLLECT_LTO_WRAPPER=/usr/lib/gcc/x86_64-pc-linux-gnu/10.2.0/lto-wrapper
Target: x86_64-pc-linux-gnu
Configured with: /build/gcc/src/gcc/configure --prefix=/usr --libdir=/usr/lib --libexecdir=/usr/lib --mandir=/usr/share/man --infodir=/usr/share/info --with-bugurl=https://bugs.archlinux.org/ --enable-languages=c,c++,ada,fortran,go,lto,objc,obj-c++,d --with-isl --with-linker-hash-style=gnu --with-system-zlib --enable-__cxa_atexit --enable-cet=auto --enable-checking=release --enable-clocale=gnu --enable-default-pie --enable-default-ssp --enable-gnu-indirect-function --enable-gnu-unique-object --enable-install-libiberty --enable-linker-build-id --enable-lto --enable-multilib --enable-plugin --enable-shared --enable-threads=posix --disable-libssp --disable-libstdcxx-pch --disable-libunwind-exceptions --disable-werror gdc_include_dir=/usr/include/dlang/gdc
Thread model: posix
Supported LTO compression algorithms: zlib zstd

$ uname -a
Linux phragyaan 5.10.7-arch1-1 #1 SMP PREEMPT Wed, 13 Jan 2021 12:02:01 +0000 x86_64 GNU/Linux
```
`/etc/makepkg.conf` - https://gist.github.com/Strykar/921e3bd01ce111670e053f9523488484
Of interest might be:
```
$ grep LDFLAGS  /etc/makepkg.conf
LDFLAGS="-Wl,-O1,--sort-common,--as-needed,-z,relro,-z,now"
```
It's a pretty standard Arch linux system.

**Comment by hoehermann** [comment]
Hm… looks like the linker settings are adjusted by the arch build environment. My changes overrule some of them. It still baffles me that the old version of the [Makefile](https://github.com/hoehermann/libpurple-signald/blob/f03cab1f6f288f6631f8a9cf9e0321d4df0773a3/Makefile) does not work, either. I do not use arch, but maybe I can find someone who can look into this.

**Comment by MarcusRiemer** [comment]
I am on a slightly less up to date standard Arch system and for me the whole thing compiles out of the box:

```
~/p/libpurple-signald ❯❯❯ gcc -v                                                                                                                                                                            master
Using built-in specs.
COLLECT_GCC=gcc
COLLECT_LTO_WRAPPER=/usr/lib/gcc/x86_64-pc-linux-gnu/10.2.0/lto-wrapper
Target: x86_64-pc-linux-gnu
Configured with: /build/gcc/src/gcc/configure --prefix=/usr --libdir=/usr/lib --libexecdir=/usr/lib --mandir=/usr/share/man --infodir=/usr/share/info --with-bugurl=https://bugs.archlinux.org/ --enable-languages=c,c++,ada,fortran,go,lto,objc,obj-c++,d --with-isl --with-linker-hash-style=gnu --with-system-zlib --enable-__cxa_atexit --enable-cet=auto --enable-checking=release --enable-clocale=gnu --enable-default-pie --enable-default-ssp --enable-gnu-indirect-function --enable-gnu-unique-object --enable-install-libiberty --enable-linker-build-id --enable-lto --enable-multilib --enable-plugin --enable-shared --enable-threads=posix --disable-libssp --disable-libstdcxx-pch --disable-libunwind-exceptions --disable-werror gdc_include_dir=/usr/include/dlang/gdc
Thread model: posix
Supported LTO compression algorithms: zlib zstd
gcc version 10.2.0 (GCC) 
~/p/libpurple-signald ❯❯❯ uname -a                                                                                                                                                                          master
Linux marcus-pc 5.10.4-arch2-1 #1 SMP PREEMPT Fri, 01 Jan 2021 05:29:53 +0000 x86_64 GNU/Linux
```

My `LDFLAGS` are slightly different though (`-z,now` seems to be missing, whatever that does):

```
~/p/libpurple-signald ❯❯❯ grep LDFLAGS  /etc/makepkg.conf                                                                                                                                             
LDFLAGS="-Wl,-O1,--sort-common,--as-needed,-z,relro"
```

I will update the system and see whether this changes anything.

**Comment by MarcusRiemer** [comment]
No issues after updating the whole system today, works on my machine™

**Comment by Strykar** [comment]
> No issues after updating the whole system today, works on my machine™

@MarcusRiemer Are you using an aur package manager or was this with `makepkg -si`?

**Comment by Strykar** [comment]
@hoehermann Some feedback from IRC, could it be [this](https://github.com/hoehermann/libpurple-signald/commit/ab105b0911c2d0a99f19a713892516c2d3ef6e10#diff-76ed074a9305c04054cdebb9e9aad2d818052b07091de1f20cad0bbac34ffb52R52):
```
20:35 <earnestly> Strykar: The build system appears to be incorrectly appending LDFLAGS after the objects: https://github.com/hoehermann/libpurple-signald/commit/ab105b0911c2d0a99f19a713892516c2d3ef6e10#diff-76ed074a9305c04054cdebb9e9aad2d818052b07091de1f20cad0bbac34ffb52R52
20:35 <phrik> Title: Compile each unit separately. · hoehermann/libpurple-signald@ab105b0 · GitHub (at github.com)
20:37 <earnestly> Perhaps even worse, they're supplying these to LD and not CC.  I don't think LD has been used like this for over a decade or more.  Make doesn't even define LD
20:39 <earnestly> Strykar: Try changing that rule to: $(CC) $^ -shared $(LDFLAGS) $(LFLIBS) -o $@
20:39 <earnestly> LDLIBS*
```
EDIT:
```
20:52 <earnestly> Strykar: Sorry: $(CC) $(LDFLAGS) -shared $^ $(LDLIBS) $@
20:53 <earnestly> This is translating the macro and rule from make's database while adding -shared.  I don't this libpurple makefile uses LDLIBS and is likely incorrectly putting -l flags in LDFLAGS
20:53 <earnestly> Strykar: You might need to then also change their LDFLAGS up above to LDLIBS (where they call pkg-config --libs)
20:58 <earnestly> The 'LIBS' suffix sort of gives it away, and why pkg-config has --libs (and even --libs-only-l, where as --libs-only-L would be flags)
```

**Comment by hoehermann** [comment]
I was not aware that my knowledge is dated and you do not invoke the linker to link binaries but use the compiler instead. I moved the definitions around as requested.

**Comment by MarcusRiemer** [comment]
> > No issues after updating the whole system today, works on my machine™
> 
> @MarcusRiemer Are you using an aur package manager or was this with `makepkg -si`?

No, I just cloned the project and ran `make clean all`.

**Comment by Strykar** [comment]
> I was not aware that my knowledge is dated and you do not invoke the linker to link binaries but use the compiler instead. I moved the definitions around as requested.

@hoehermann I pasted the chat in a haste, as this was outside the scope of my abilities to troubleshoot.. should've kept it relevant. Your change fixes it and it builds fine now, yay!! Thank you!

@MarcusRiemer Makes sense, Arch's aur / `/etc/makepkg.conf` were not involved.

**Comment by hoehermann** [comment]
@Strykar No problem. My post was not meant to be passive-aggressive. I honestly did not know I was using a style which is so old it is now frowned upon (e.g. https://stackoverflow.com/questions/41372118/whats-the-relationship-between-gcc-linking-and-ld-linking/#41372161).

All in all, it compiles for you, it works for and @demure. So I am closing this issue.
